br/pkg/backup/prepare_snap/stream.go

```diff
@@ -0,0 +1,209 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package preparesnap
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/pingcap/errors"
+	brpb "github.com/pingcap/kvproto/pkg/brpb"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/br/pkg/logutil"
+	"github.com/pingcap/tidb/br/pkg/utils"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type eventType int
+
+const (
+	eventMiscErr eventType = iota
+	eventWaitApplyDone
+)
+
+type event struct {
+	ty eventType
+
+	storeID uint64
+	err     error
+	region  *metapb.Region
+}
+
+func (e event) String() string {
+	return fmt.Sprintf("Event(Type: %v, StoreID: %v, Error: %v, Region: %v)", e.ty, e.storeID, e.err, e.region)
+}
+
+type prepareStream struct {
+	storeID       uint64
+	cli           PrepareClient
+	leaseDuration time.Duration
+
+	output       chan<- event
+	serverStream <-chan utils.Result[*brpb.PrepareSnapshotBackupResponse]
+
+	clientLoopHandle *errgroup.Group
+	stopBgTasks      context.CancelFunc
+}
+
+// InitConn initializes the connection to the stream (i.e. "active" the stream).
+//
+// Before calling this, make sure you have filled the store ID and output channel.
+//
+// Once this has been called, them **should not be changed** any more.
+func (p *prepareStream) InitConn(ctx context.Context, cli PrepareClient) error {
+	p.cli = cli
+	p.clientLoopHandle, ctx = errgroup.WithContext(ctx)
+	ctx, p.stopBgTasks = context.WithCancel(ctx)
+	return p.GoLeaseLoop(ctx, p.leaseDuration)
+}
+
+func (p *prepareStream) Finalize(ctx context.Context) error {
+	log.Info("shutting down", zap.Uint64("store", p.storeID))
+	return p.stopClientLoop(ctx)
+}
+
+func (p *prepareStream) GoLeaseLoop(ctx context.Context, dur time.Duration) error {
+	err := p.cli.Send(&brpb.PrepareSnapshotBackupRequest{
+		Ty:             brpb.PrepareSnapshotBackupRequestType_UpdateLease,
+		LeaseInSeconds: uint64(dur.Seconds()),
+	})
+	if err != nil {
+		return errors.Annotate(err, "failed to initialize the lease")
+	}
+	msg, err := p.cli.Recv()
+	if err != nil {
+		return errors.Annotate(err, "failed to recv the initialize lease result")
+	}
+	if msg.Ty != brpb.PrepareSnapshotBackupEventType_UpdateLeaseResult {
+		return errors.Errorf("unexpected type of response during creating lease loop: it is %s", msg.Ty)
+	}
+	p.serverStream = utils.AsyncStreamBy(p.cli.Recv)
+	p.clientLoopHandle.Go(func() error { return p.clientLoop(ctx, dur) })
+	return nil
+}
+
+func (p *prepareStream) onResponse(ctx context.Context, res utils.Result[*brpb.PrepareSnapshotBackupResponse]) error {
+	if err := res.Err; err != nil {
+		return err
+	}
+	resp := res.Item
+	logutil.CL(ctx).Debug("received response", zap.Stringer("resp", resp))
+	evt, ok := p.convertToEvent(resp)
+	if ok {
```

Comment:
Do we need handle `else` herer?

Reply:
Some of events may not need to be delivered to upper layer. (A unknown message will generate an error instead of being dropped.)
Perhaps the name `ok` is misleading.